mlir/lib/Quantum/Transforms/ChainedSelfInversePatterns.cpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#define DEBUG_TYPE "chained-self-inverse"
+
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/Errc.h"
+
+#include "Quantum/IR/QuantumOps.h"
+#include "Quantum/Transforms/Patterns.h"
+
+using llvm::dbgs;
+using namespace mlir;
+using namespace catalyst;
+using namespace catalyst::quantum;
+
+namespace {
+
+llvm::SmallVector<StringRef> SelfInverseGates{
```

Comment:
To reduce the number of constexpr lists of gates in Catalyst and provide a better control over which gates can be cancelled at compile-time, does it make sense to move this list to the frontend?

Reply:
Interesting idea, that would give the user control over, say only cancelling Hadamards and leaving Paulis in, if they wish to do so. I think this is a very good suggestion, as the more control the UI can give the user the better. I will think about how to implement this. 

An immediate design I can think of is to make the pass take in a comma-separated string option that indicates what gates to cancel (e.g. -gates="Hadamard,PauliZ,CNOT"), and default to all the gates listed here if nothing is provided (this should be easily doable in the tablegen). Then the pass itself does not have to keep this contexpr list and only needs to do some trivial parsing on this string. 

The frontend UI can take in another option that controls this, i.e. `@catalyst.cancel_inverses(target_gates="Hadamard,PauliZ,CNOT")`. Thoughts @josh146 ?

Reply:
@paul0403 We should absolutely move this list to the frontend, although from a Python UI point of view I would suggest a UI following this: https://app.shortcut.com/xanaduai/epic/66937?group_by=none&ct_workflow=all&cf_workflow=500000005

That is, providing a list of gate strings (or also a list of PL gate classes if that makes sense).

I would also consider here; is this out of scope for the current PR, and worth making a following on story/PR for? You'll have more context here, but typically the more concise/scoped we can make each PR the better/faster review will be